### PR TITLE
[Keyboard] Kinesis + Pro Micro

### PR DESCRIPTION
**Pull request by nguyenvietyen**
This is my first work ever with QMK - I'm not intimate yet with all details and considerations.

This PR comprises two parts:
- mod to retrofit the Kinesis Advantage keyboard with a Arduine Pro Micro (clone). 
- a Dvorak keymap that is based on the Kinesis default. The ESC and Caps lock keys are swapped. 

I'd appreciate feedback, particularly on:
- ~~how to implement a faster debounce strategy for reducing lag~~ DONE.
- ~~whether setting the pinouts is necessary too in `nguyenvietyen.c`~~
- general code conventions and potential cleanups 

Thank you!

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [X] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kinesis/config.h [hunk 1]
```diff
@@ -139,5 +139,10 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #ifdef SUBPROJECT_kint2pp
     #include "kint2pp/config.h"
 #endif
+#ifdef SUBPROJECT_nguyenvietyen
```
Comment (drashna) [inline keyboards/kinesis/config.h:142]: These aren't needed, actually, as they're automatically included.

File: keyboards/kinesis/keymaps/dvorak_nguyenvietyen/config.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef CONFIG_USER_H
+#define CONFIG_USER_H
+
+#include "../../config.h"
+
+// place overrides here
+
+#endif
```
Comment (drashna) [inline keyboards/kinesis/keymaps/dvorak_nguyenvietyen/config.h:8]: This entire file can be removed.
```suggestion
```

File: keyboards/kinesis/keymaps/dvorak_nguyenvietyen/keymap.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  return true;
+}
+
+void led_set_user(uint8_t usb_led) {
+
+}
```
Comment (drashna) [inline keyboards/kinesis/keymaps/dvorak_nguyenvietyen/keymap.c:46]:
```suggestion
```

File: keyboards/kinesis/rules.mk [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # MCU name
-MCU = at90usb1286
+MCU = atmega32u4
```
Comment (noroadsleft) [inline keyboards/kinesis/rules.mk:2]: This file should remain unchanged, and you should override the necessary rules in `keyboards/kinesis/nguyenvietyen/rules.mk` instead.

```suggestion
MCU = at90usb1286
```

File: keyboards/kinesis/rules.mk [hunk 5]
```diff
@@ -9,7 +9,7 @@ MCU = at90usb1286
 #   QMK DFU      qmk-dfu
 #   ATmega32A    bootloadHID
 #   ATmega328P   USBasp
-BOOTLOADER = halfkay
+BOOTLOADER = caterina
```
Comment (noroadsleft) [inline keyboards/kinesis/rules.mk:12]:
```suggestion
BOOTLOADER = halfkay
```

File: keyboards/kinesis/nguyenvietyen/rules.mk [hunk 6]
```diff
@@ -0,0 +1,4 @@
+CUSTOM_MATRIX = yes # need to do our own thing with the matrix
```
Comment (noroadsleft) [inline keyboards/kinesis/nguyenvietyen/rules.mk:1]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = caterina

CUSTOM_MATRIX = yes # need to do our own thing with the matrix
```

File: keyboards/kinesis/nguyenvietyen/README.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+   FS1	11	30	ALE
+ CLOCK	12	29	PSEN
+  DATA	13	28	r8
+	    14	27	r7
+   FS3	15	26	r6
+	    16	25	r5
+   FS2	17	24	r4
+ XTAL1	18	23	r3
+ XTAL2	19	22	r2
+   GND	20	21	r1
```
Comment (noroadsleft) [inline keyboards/kinesis/nguyenvietyen/README.md:70]: Alignment fixes.

```suggestion
   DL2   1  40  VCC
   DR1   2  39  A
   DR2   3  38  B
   DL1   4  37  C
   KPD   5  36  G
   PGM   6  35
   SCL   7  34
   SDA   8  33
   RST   9  32  BUZZ
    WP  10  31  EA
   FS1  11  30  ALE
 CLOCK  12  29  PSEN
  DATA  13  28  r8
        14  27  r7
   FS3  15  26  r6
        16  25  r5
   FS2  17  24  r4
 XTAL1  18  23  r3
 XTAL2  19  22  r2
   GND  20  21  r1
```

File: keyboards/kinesis/nguyenvietyen/README.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+```
+(this is copied from [@wjanssens](https://raw.githubusercontent.com/wjanssens/tmk_keyboard/master/keyboard/kinesis/doc/readme.txt))
+
+## Pin Mapping to Pro Micro
+
+![Pro Micro Pinout](doc/pro_micro_pinout.jpg)
```
Comment (noroadsleft) [inline keyboards/kinesis/nguyenvietyen/README.md:76]: This image link doesn't work.

**Comment by nguyenvietyen** [comment]
Thank you all for the constructive feedback! 

I'll do the cleanups and prepare this PR for merging.

File: keyboards/kinesis/config.h (hunk 1 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/config.h]: Great, I've removed them.

File: keyboards/kinesis/keymaps/dvorak_nguyenvietyen/config.h (hunk 2 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/keymaps/dvorak_nguyenvietyen/config.h]: Done 👍

File: keyboards/kinesis/keymaps/dvorak_nguyenvietyen/keymap.c (hunk 3 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/keymaps/dvorak_nguyenvietyen/keymap.c]: Done 👍

File: keyboards/kinesis/rules.mk (hunk 4 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/rules.mk]: Done 👍

File: keyboards/kinesis/rules.mk (hunk 5 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/rules.mk]: Done 👍

File: keyboards/kinesis/nguyenvietyen/rules.mk (hunk 6 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/rules.mk:1]: Done 👍

File: keyboards/kinesis/nguyenvietyen/README.md (hunk 7 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/README.md]: Done 👍

File: keyboards/kinesis/nguyenvietyen/README.md (hunk 8 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/README.md]: Looks like .jpg is in gitignore. I've removed the image in favor of a table.

**Comment by nguyenvietyen** [comment]
Cleanups are done. This PR should be ready for merging, unless further feedback arises.

File: keyboards/kinesis/nguyenvietyen/config.h [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * ROWS: AVR pins used for rows, top to bottom
+ * DIODE_DIRECTION: COL2ROW = COL = Anode (+), ROW = Cathode (-, marked on diode)
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ *
+ */
+// Passed through the port multipler, so 4 pins =16
+#define MATRIX_ROW_PINS \
+    { D0, D1, D2, D3 }
+#define MATRIX_COL_PINS \
+    { B6, B2, B3, B1, F7, F6, F5, F4 }
```
Comment (fauxpark) [inline keyboards/kinesis/nguyenvietyen/config.h:25]:
```suggestion
#define MATRIX_ROW_PINS { D0, D1, D2, D3 }
#define MATRIX_COL_PINS { B6, B2, B3, B1, F7, F6, F5, F4 }
```

File: keyboards/kinesis/keymaps/dvorak_nguyenvietyen/rules.mk [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+NKRO_ENABLE = yes           # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+
```
Comment (fauxpark) [inline keyboards/kinesis/keymaps/dvorak_nguyenvietyen/rules.mk:18]: This file should be deleted.

File: keyboards/kinesis/nguyenvietyen/rules.mk [hunk 11]
```diff
@@ -0,0 +1,13 @@
+CUSTOM_MATRIX = lite
+
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = caterina
+
+# Project specific files
```
Comment (fauxpark) [inline keyboards/kinesis/nguyenvietyen/rules.mk:9]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no
COMMAND_ENABLE = yes
NKRO_ENABLE = yes
CUSTOM_MATRIX = lite

# Project specific files
```

File: keyboards/kinesis/nguyenvietyen/rules.mk [hunk 12]
```diff
@@ -0,0 +1,13 @@
+CUSTOM_MATRIX = lite
+
```
Comment (fauxpark) [inline keyboards/kinesis/nguyenvietyen/rules.mk:2]:
```suggestion
```

File: keyboards/kinesis/nguyenvietyen/matrix.c [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        bool row_has_changed = current_matrix[i] != row;
+        matrix_has_changed |= row_has_changed;
+        current_matrix[i] = row;
+    }
+
+    return matrix_has_changed;
+}
+
+static matrix_row_t read_row(uint8_t row) {
+    _delay_us(30);  // without this wait read unstable value.
```
Comment (fauxpark) [inline keyboards/kinesis/nguyenvietyen/matrix.c:54]:
```suggestion
    matrix_io_delay();  // without this wait read unstable value.
```

File: keyboards/kinesis/nguyenvietyen/rules.mk (hunk 12 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/rules.mk:2]: According to https://beta.docs.qmk.fm/using-qmk/hardware-features/custom_matrix, this is needed. Is the doc perhaps incorrect?

File: keyboards/kinesis/keymaps/dvorak_nguyenvietyen/rules.mk (hunk 10 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/keymaps/dvorak_nguyenvietyen/rules.mk:18]: Done :)

File: keyboards/kinesis/nguyenvietyen/rules.mk (hunk 12 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/rules.mk:2]: Oh wait, I missed now. Nvm.

File: keyboards/kinesis/nguyenvietyen/matrix.c (hunk 13 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/matrix.c:54]: Done 👍

File: keyboards/kinesis/nguyenvietyen/matrix.c [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+static void         unselect_rows(void);
+static void         select_rows(uint8_t row);
+
+void matrix_init_custom(void) {
+    // output low (leds)
+    setPinOutput(D7);  // Keypad LED
+    setPinOutput(C6);  // ScrLock LED
+    setPinOutput(D4);  // NumLock LED
+    setPinOutput(E6);  // CapsLock LED
+
```
Comment (fauxpark) [inline keyboards/kinesis/nguyenvietyen/matrix.c:13]:
```suggestion
```

File: keyboards/kinesis/nguyenvietyen/matrix.c [hunk 15]
```diff
@@ -0,0 +1,72 @@
+#include "nguyenvietyen.h"
```
[the remark this answers is not in the document]
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/matrix.c]: Functions are used that are not in `matrix.h`, instead I included "quantum.h" instead.

File: keyboards/kinesis/nguyenvietyen/matrix.c [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+static matrix_row_t read_row(uint8_t row);
+static void         unselect_rows(void);
+static void         select_rows(uint8_t row);
```
[the remark this answers is not in the document]
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/matrix.c]: Done 👍

File: keyboards/kinesis/nguyenvietyen/matrix.c (hunk 14 above)
Comment (nguyenvietyen) [reply keyboards/kinesis/nguyenvietyen/matrix.c]: Done 👍

File: keyboards/kinesis/nguyenvietyen/matrix.c [hunk 17]
```diff
@@ -0,0 +1,62 @@
+#include "quantum.h"
```
Comment (fauxpark) [inline keyboards/kinesis/nguyenvietyen/matrix.c:1]:
```suggestion
#include "matrix.h"

#include "quantum.h"
```

**Comment by nguyenvietyen** [comment]
I just discovered a minor led activation bug when hooked to my docking station while the laptop is disconnected.

Please hold off merging until I push the fix.

**Comment by nguyenvietyen** [comment]
> I just discovered a minor led activation bug when hooked to my docking station while the laptop is disconnected.
> 
> Please hold off merging until I push the fix.

I've investigated it a bit. The Docking station is practically a USB hub. When the laptop gets disconnected, it looks like QMK is sending 0 to all led outputs. Due to the Kinesis wiring, all keyboard LEDs are switched on by that.

I'll look further for a solution.

So this doesn't block merging, but it would be nice to address it as part of this PR.

**Comment by nguyenvietyen** [comment]
> > I just discovered a minor led activation bug when hooked to my docking station while the laptop is disconnected.
> > Please hold off merging until I push the fix.
> 
> I've investigated it a bit. The Docking station is practically a USB hub. When the laptop gets disconnected, it looks like QMK is sending 0 to all led outputs. Due to the Kinesis wiring, all keyboard LEDs are switched on by that.
> 
> I'll look further for a solution.
> 
> So this doesn't block merging, but it would be nice to address it as part of this PR.

I've just fixed the LED activation bug.

The PR is ready again for review (and merging).

File: keyboards/kinesis/nguyenvietyen/nguyenvietyen.c [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    setPinOutput(D7);  // Keypad LED
+    setPinOutput(C6);  // ScrLock LED
+    setPinOutput(D4);  // NumLock LED
+    setPinOutput(E6);  // CapsLock LED
+
+    // turn all LEDs off by default
+    writePin(D7, 1);
+    writePin(C6, 1);
+    writePin(D4, 1);
+    writePin(E6, 1);
```
Comment (fauxpark) [inline keyboards/kinesis/nguyenvietyen/nguyenvietyen.c:19]:
```suggestion
    writePinHigh(D7);
    writePinHigh(C6);
    writePinHigh(D4);
    writePinHigh(E6);
```

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**